python/metatensor-operations/metatensor/operations/_testing/_grad.py

```diff
@@ -0,0 +1,78 @@
+from typing import Callable, Union
+
+import numpy as np
+from numpy.testing import assert_allclose
+
+from .. import _dispatch
+from .._classes import TensorMap
+from .._dispatch import TorchTensor
+
+
+def finite_differences(
+    function: Callable[[Union[np.ndarray, TorchTensor], bool], TensorMap],
```

Comment:
ah, we potentially need multiple inputs:

```suggestion
    function: Callable[[TensorMap, TensorMap], TensorMap],
```

and give the function two different random inputs (the function is free to ignore one of them for e.g. `pow`). Ideally two inputs with different set of gradient samples would be nice, but we can start with identical set of gradients.

Reply:
I will try it out to see how well this works.

Reply:
I am not sure about the two TensorMap's as input as I don't see much what kind of bugs this could catch which is not covered by the current version and it makes everything a bit more cumbersome. The idea in the current implementation is that you still can create two different sets of gradients but that inherit from the same input, so the finite difference test is easier to perform. In the test for add we first create two different gradients with `cartesian_cubic` and `cartesian_linear` to then pass it to the `add`. So we are testing  this `add(f(X), g(X)`. I am not sure what kind of bug `add(f(X), g(Y))` would catch in addition.

What I see is missing in the current design is a way to also test operations that affect the samples as sort and reduce over samples does.